include/seqan3/io/alignment_file/format_bam.hpp

```diff
@@ -849,24 +851,37 @@ class alignment_file_output_format<format_bam> : alignment_file_output_format<fo
             // ---------------------------------------------------------------------
             // Writing the Record
             // ---------------------------------------------------------------------
+            int32_t ref_length{};
 
-            // compute possible distance from alignment end to sequence end
-            // which indicates soft clipping at the end.
-            // This should be replaced by a free count_gaps function for
-            // aligned sequences which is more efficient if possible.
-            auto off_end{std::ranges::distance(seq) - offset};
+            // if alignment is non-empty, replace cigar_vector.
+            // else, compute the ref_length from given cigar_vector which is needed to fill field `bin`.
+            if (!std::ranges::empty(get<0>(align)) && !std::ranges::empty(get<1>(align)))
+            {
+                ref_length = std::ranges::distance(get<1>(align));
 
-            for (auto chr : get<1>(align))
-                if (chr == gap{})
-                    ++off_end;
+                // compute possible distance from alignment end to sequence end
+                // which indicates soft clipping at the end.
+                // This should be replaced by a free count_gaps function for
+                // aligned sequences which is more efficient if possible.
+                auto off_end{std::ranges::distance(seq) - offset};
```

Comment:
Is this `auto` a signed type? If not you should have an assertion below that `off_end -= ref_length;`cannot become negative. 

Anyway I would not use auto here, as the resulting type from the difference cannot be easily seen.